x/gov/keeper/proposal.go

```diff
@@ -95,6 +95,38 @@ func (keeper Keeper) SubmitProposal(ctx sdk.Context, messages []sdk.Msg, metadat
 	return proposal, nil
 }
 
+// CancelProposal will cancel proposal before the voting period ends
+func (keeper Keeper) CancelProposal(ctx sdk.Context, proposalID uint64, proposer string) error {
+	proposal, ok := keeper.GetProposal(ctx, proposalID)
+	if !ok {
+		return sdkerrors.Wrapf(types.ErrProposalNotFound, "proposal is not found with %d id", proposalID)
+	}
+
+	// Check creator of the proposal
+	if proposal.Proposer != proposer {
+		return sdkerrors.Wrapf(types.ErrInvalidProposer, "invalid proposer %s", proposer)
+	}
+
+	// Check if proposal is active or not
+	if (proposal.Status != v1.StatusDepositPeriod) && (proposal.Status != v1.StatusVotingPeriod) {
```

Comment:
this cancels a proposal that has entered the voting period. the issue only talks about deposit period. Voting period, imo should not be cancellable because its an escape for no w veto burning.

Reply:
I think you should be able to cancel *only* during the voting period and the deposit is automatically burned
If you want to cancel during the deposit period it's maybe fine to let them keep what balance has been added?

Reply:
@marbar3778  @hxrts  Now we cancel the proposal before the voting period ends (in the deposit period and voting period also), 
We are sending the proposal minimum deposits from the proposal to the community pool and refunding the remaining deposits to depositors.

Reply:
ok I think that's probably fine

Reply:
> @marbar3778 @hxrts Now we cancel the proposal before the voting period ends (in the deposit period and voting period also),
> We are sending the proposal minimum deposits from the proposal to the community pool and refunding the remaining deposits to depositors.

**If we cancel the proposal in the voting period there is a chance of escape for no w veto burning.**
https://github.com/cosmos/cosmos-sdk/pull/13010#discussion_r963469884 

cc: @alexanderbez

Reply:
May be we should not allow the proposal to be cancelled for which the tally can be `NoWithVeto`

Reply:
I don't have a strong opinion but we can consider this, I cannot think of any downsides with this (and may be force this only if the quorum is reached?)

Reply:
ok so my understanding is the current implementation allows cancellation ONLY during the voting period and the deposit it sent to the community pool if cancellation is successful. there is no escape of now_with_veto here so this works just fine.

I don't think it's necessary to have cancellation during the deposit period. It's not totally clear what should happen to the funds, and with the voting period-only implementation proposers can either let the proposal expire without adding the remaining deposit, or cancel once it enters the voting period.

Reply:
My 2 cents and this seems to be inline with what @gsk967 and @hxrts have pointed out:

Cancellation during:

* Proposal in **VOTING** period -> send all deposits to community pool and only allow proposer to cancel
* Proposal in **DEPOSIT** period -> send deposits back to all depositors

Reply:
@alexanderbez I don't think people would be interested to cancel any proposals if all the deposits goes to community pool. That's why we decided to use min-proposer-deposit and return remaining deposits.

But after reading through different POVs and confusion, may be we can just allow edit proposals in deposit period and withdraw cancel proposal feature completely. This covers majority of the cases that we initially discussed about ie., any mistakes in the proposal/upgrade heights.

Reply:
> But after reading through different POVs and confusion, may be we can just allow edit proposals in deposit period and withdraw cancel proposal feature completely. 

Yeah that makes sense for proposals that are still in deposit period, but if you do that, then 

1. you also need to allow the ability for depositors to withdraw their deposit if they do not agree with the edit(s)
2. you must also still allow proposals in _voting_ period to be cancelable

Reply:
Agree with 1 but for 2, with cancelling we are not taking any consent from the depositors. If the proposer just wants to cancel, moving all deposits to community pool might not be interesting for depositors.

Reply:
So cancelling a prop in VP, you want to send all deposits back to depositors and send the proposer's deposit to the community pool?

Reply:
I would advise for that.

Reply:
I think all deposits need to go to community pool, otherwise proposers can use a secondary account to make most of their deposit risk-free.

Reply:
I agree with @anilCSE . Proposals should be immutable
1. If I supported a proposal, and contributed to a deposit, I don't want to be surprised that someone edited it.
2. As @alexanderbez noted, and discussed previously, we have to make sure that we can cancel a proposal during the voting period, because sometimes we find a bug only during the voting period. I would like to second my proposal for cancelling process during the voting (or later:
    1. Once proposal passed it must not be withdrawn (in emergency we can always do a coordinated hardfork, or if there is still time, we can make a new proposals to cancel the previous one -> @anilCSE  also suggested that)
    2.  to avoid manipulation, cancelling should be possible strictly before the VP end (eg, 10% before - can be a gov param) https://github.com/cosmos/cosmos-sdk/discussions/11130#discussioncomment-3136147
    3. There has to be some cost of cancelling proposal, otherwise we face spam issues (as also suggested by @ValarDragon  https://github.com/cosmos/cosmos-sdk/discussions/11130#discussioncomment-2511040  and @hxrts ). If we don't want to send 100% to the community pool, then how about 50% (can be a gov param)?

Reply:
So team, did we agree on an approach here? I don't want to keep the PR looming for too long in interest of avoiding continual merge conflict resolution burden for the PR author 🙏

Reply:
Maybe we should have a breakout call? Seams like an important design decision. Otherwise we can try to handle it during the community / architecture call.

Reply:
Talked with @hxrts about the point ii. from https://github.com/cosmos/cosmos-sdk/pull/13010#discussion_r967820511. He convinced me that limiting cancellation to x% time of the proposal length is not needed - and the cost param is enough  (point iii.).

Reply:
@robert-zaremba wdyt about https://github.com/cosmos/cosmos-sdk/pull/13010#discussion_r1001262872